river/neighbors/base_neighbors.py

```diff
@@ -1,233 +1,222 @@
+    Updates are by default stored by the FIFO (first in first out) method,
+    which means that when the size limit is reached, old samples are dumped to
+    give room for new samples. This is circular, meaning that older points
+    are dumped first. This also gives the implementation a temporal aspect,
+    because older samples are replaced with newer ones. However, if
+    min_distance_keep is set to be != 0 every sample given to learn may not be
+    added to the window. When using this class, the "x" array for the window
+    can be a set of any (x, y, <attributes>) and the class always assumes x[0]
+    to be features x, x[1] to be y, and x[2] .. x[n-2] where N is the length of
+    the array to be any set of custom features added by the class. After the
+    metric is run, the distance should always be last in the array, x[-1]
+    or x[n-1] given an array of length n.
 
     """
 
-    def __init__(self, window_size: int = 1000):
+    def __init__(self, window_size, p, min_distance_keep=0.05):
         self.window_size = window_size
-        self._n_features: int = -1
-        self._n_targets: int = -1
-        self._size: int = 0
-        self._next_insert: int = 0
-        self._oldest: int = 0
-        self._imask: np.ndarray
-        self._X: np.ndarray
-        self._y: typing.List
-        self._is_initialized: bool = False
-
-    def _configure(self):
-        # Binary instance mask to filter data in the buffer
-        self._imask = np.zeros(self.window_size, dtype=bool)
-        self._X = np.zeros((self.window_size, self._n_features))
-        self._y = [None for _ in range(self.window_size)]
-        self._is_initialized = True
-
-    def reset(self):
-        """Reset the sliding window."""
-        self._n_features = -1
-        self._n_targets = -1
-        self._size = 0
-        self._next_insert = 0
-        self._oldest = 0
-        self._imask = None
-        self._X = None
-        self._y = None
-        self._is_initialized = False
-
-        return self
-
-    def append(self, x: np.ndarray, y: base.typing.Target) -> "KNeighborsBuffer":
-        """Add a (single) sample to the sample window.
-
-        x
-            1D-array of feature for a single sample.
-
-        y
-            The target data for a single sample.
-
-        Raises
-        ------
-        ValueError: If at any moment, a sample with a different number of
-        attributes than that of the n_attributes parameter is passed, a
-        ValueError is raised.
+        self.p = p
+        self.window = collections.deque(maxlen=window_size)
 
-        TypeError: If the buffer type is altered by the user, or is not
-        correctly initialized, a TypeError may be raised.
+        # A minimum distance (similarity) to determine adding to window
+        # The model will perform better with a more diverse window
+        self.min_distance_keep = min_distance_keep
 
+    def append(self, x, y=None, extra: [Tuple, list] = None):
```

Comment:
Regarding passing parameters to the dist function, I believe the way to go is using `functools.partial`. We have used that in other situations in our codebase.

For example:

```python
dist_func=functools.partial(utils.minkowski_distance, p=2)
```

Reply:
By default:

`dist_func=None`

and then 

```python
# __init__
self.dist_func = dist_func if dist_func is not None else functools.partial(utils.minkowski_distance, p=2)

```

Reply:
> In this case, the p parameter and Minkowski distance would be dropped too, and replaced by something like dist_func. Such distance function, besides calculating the distances (duh), would also be in task of selecting what to use from the item.

I also already added this - do you want to review the current version?

Reply:
I'm not familiar with functools partial, and I'm not totally following the "allow any kind of thing" because it would be different than the other river classes. I think this is probably a good splitting point for @MaxHalford to take over and do tweaks to his liking.

Reply:
I see. How would that account for:

1. handling custom values of p for the minkowski metric set by the user, and
2. allowing the user to change it on the fly easily OR set the default again easily?

What benefit does functools partial have over just saving a pointer to the function and using it?